compiler/rustc_metadata/src/rmeta/encoder.rs

```diff
@@ -485,18 +488,22 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
             })
             .map(|(_, source_file)| {
                 let mut adapted = match source_file.name {
-                    // This path of this SourceFile has been modified by
-                    // path-remapping, so we use it verbatim (and avoid
-                    // cloning the whole map in the process).
-                    _ if source_file.name_was_remapped => source_file.clone(),
-
-                    // Otherwise expand all paths to absolute paths because
-                    // any relative paths are potentially relative to a
-                    // wrong directory.
                     FileName::Real(ref name) => {
-                        let name = name.stable_name();
+                        // Expand all local paths to absolute paths because
+                        // any relative paths are potentially relative to a
+                        // wrong directory.
                         let mut adapted = (**source_file).clone();
-                        adapted.name = Path::new(&working_dir).join(name).into();
+                        adapted.name = match name {
+                            RealFileName::Named(local_path) => {
+                                Path::new(&working_dir).join(local_path).into()
+                            }
+                            RealFileName::Virtualized { local_path, virtual_name } => {
+                                FileName::Real(RealFileName::Virtualized {
+                                    local_path: Path::new(&working_dir).join(local_path),
```

Comment:
The local path of a file must not end up in metadata. That's one of the requirements for path remapping (because of reproducible builds and privacy issues). 

If you look carefully, the existing implementation "flattens" the filename from `RealFileName::Devirtualized` into `RealFileName::Named` in line 499. Similarly, in the new version we need to map any `RealFileName::Virtualized` to a `RealFileName::Named` leaving only the `virtual_name` in there.

Reply:
The comment here is a bit misleading because it's not really "expand all paths to absolute paths". I just realised that working_dir could've also been remapped hence not absolute. I'll change it. 

If we were to remap all `RealFileName::Virtualized` to a `RealFileName::Named`, I suppose we would only need to deal with `Named` variant in `rustc_metadata/src/rmeta/decoder.rs:imported_source_files()`? Although this would make preventing double remapping difficult (probably impossible) without retaining the `name_was_remapped` field in `rustc_span::SourceFile`

Reply:
I'm not quite happy with the current `RealFileName` variants. The name of the `Named` variant does not tell you what it is for and is potentially confusing (if there is a `Named` variant then why isn't there an `Unnamed` variant, etc...). The `Named` variant can also be either (1) the local path or (2) the stable name, depending on whether remapping was involved or not. And we actually don't always have a name that is really "stable" in the sense of being system independent -- even remapping does not guarantee that the result is stable in that sense. 

So what a `RealFileName` actually can be is:

- The local path to a file
- A remapped path containing the local path and the result of remapping
- A path from upstream crate metadata which can either be 
   - an absolute local path or 
   - a remapped path without the local part

And depending on the situation we want to use the "most stable" path (for emitting to output artefacts) or the path that actually allows you to load something from disk (parsing & error messages).

I think I would suggest the following:

- Rename `RealFileName::Named` to `RealFileName::LocalPath` -- because that's what it is, right?
- Rename ` RealFileName::Virtualized` to `RealFileName::Remapped` so that the naming does not clash with "virtual files" generated by macro expansion.
- Make `local_path` in `RealFileName::Remapped` an `Option<_>` so that we can explicitly "erase" the local portion of the remapped path during export.
- Manually implement `Encodable` for `RealFileName` and panic if we encounter a `local_path` that wasn't erased (i.e. `assert!(local_path.is_none())`).
- Maybe rename the `stable_name` method to something like `most_stable_name` and documentation that it will fall back on the local path if that's the only thing available.
- Use `RealFileName` for the `working_dir` field in `Session` to make things explicit there too.

Reply:
How does that sound to you?

Reply:
Yup that sounds a lot cleaner

Reply:
Another approach to making `local_path` in `RealFileName::Remapped` an `Option<_>` is to introduce another variant `FileName::External`, which takes two values: the path and a boolean indicating if it has been remapped already. The boolean is needed so that already remapped paths from the metadata of another crate won't be remapped again (although those haven't been remapped will be, otherwise we can't do anything about `$HOME/.cargo/registry/...` paths everywhere). A `FileName::Real` is turned into `FileName::External` in `rustc_metadata::rmeta::encoder`.

This makes an explicit distinction between self-generated and imported metadata and prevents needing all those changes to librustdoc in https://github.com/rust-lang/rust/pull/83813/commits/fdcae961afc0369dc51438bfe2ec2b516a83673d